pkg/v1/tarball/write.go

```diff
@@ -29,83 +29,112 @@ import (
+func WriteToFile(p string, ref name.Reference, img v1.Image, updates chan<- v1.Update) error {
 	w, err := os.Create(p)
 	if err != nil {
 		return err
 	}
 	defer w.Close()
 
-	return Write(ref, img, w)
+	return Write(ref, img, w, updates)
 }
 
 // MultiWriteToFile writes in the compressed format to a tarball, on disk.
 // This is just syntactic sugar wrapping tarball.MultiWrite with a new file.
-func MultiWriteToFile(p string, tagToImage map[name.Tag]v1.Image) error {
+func MultiWriteToFile(p string, tagToImage map[name.Tag]v1.Image, updates chan<- v1.Update) error {
 	refToImage := make(map[name.Reference]v1.Image, len(tagToImage))
 	for i, d := range tagToImage {
 		refToImage[i] = d
 	}
-	return MultiRefWriteToFile(p, refToImage)
+	return MultiRefWriteToFile(p, refToImage, updates)
 }
 
 // MultiRefWriteToFile writes in the compressed format to a tarball, on disk.
 // This is just syntactic sugar wrapping tarball.MultiRefWrite with a new file.
-func MultiRefWriteToFile(p string, refToImage map[name.Reference]v1.Image) error {
+func MultiRefWriteToFile(p string, refToImage map[name.Reference]v1.Image, updates chan<- v1.Update) error {
 	w, err := os.Create(p)
 	if err != nil {
 		return err
 	}
 	defer w.Close()
 
-	return MultiRefWrite(refToImage, w)
+	return MultiRefWrite(refToImage, w, updates)
 }
 
 // Write is a wrapper to write a single image and tag to a tarball.
-func Write(ref name.Reference, img v1.Image, w io.Writer) error {
-	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w)
+func Write(ref name.Reference, img v1.Image, w io.Writer, updates chan<- v1.Update) error {
+	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w, updates)
 }
 
 // MultiWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer) error {
+func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer, updates chan<- v1.Update) error {
 	refToImage := make(map[name.Reference]v1.Image, len(tagToImage))
 	for i, d := range tagToImage {
 		refToImage[i] = d
 	}
-	return MultiRefWrite(refToImage, w)
+	return MultiRefWrite(refToImage, w, updates)
 }
 
 // MultiRefWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer) error {
+func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer, updates chan<- v1.Update) error {
```

Comment:
This is a breaking change for callers, which I'd like to avoid if possible. We've been using [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) for this, so I'd probably change this to something like:

```
func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer, opts... Option) error {
```

and add a `WithProgress` option (or something, names are hard).

You can see how we've done this with remote: https://godoc.org/github.com/google/go-containerregistry/pkg/v1/remote#Option

Reply:
> I'd probably change this to something like

I am ok with variadic options, makes lots of sense. Not only does it make it non-breaking, but makes it easy to expand going forward.

> and add a WithProgress option 

I am not overly enamoured of functional options (although I give credit to Dave Cheney; he really laid out the challenges and the solutions nicely in that blog post). I get what they are trying to solve, and do so quite well, but they can lead to additional complexity.

* It is harder to visualize and reason about functions as options that just straight parameters or config structs, and I really believe builder complexity is the most expensive cost
* It requires the passed object to expose a lot of members that might remain private (see Cheney's examples where he has to modify `srv.timeout` or `srv.listener`), or expose a potential slew of new methods on the object (e.g. `srv.SetTimeout()` or `srv.SetListener()`; they also might be called later, when it makes no sense (a point he raises early on the post). That is what config opts are good for.

That having been said (i.e. discussing it, maybe I will get some new insights from someone smart), if you already are using them, we can just adopt it here. I will modify, let us see.

Reply:
I will say one thing; it makes the tests unchanged, which is nice.

Reply:
I updated it to use variadic function opts. Take a look now and see.

Reply:
> I am not overly enamoured of functional options

Yeah, I have mixed feelings about them too :/ we might overuse them, but they're nice to avoid breaking changes!